### DISCUSSION TITLE
Add per-Collection (or per-Environment) HTTP(S) proxy setting

### DISCUSSION THREAD

**Discussion by cpick** [Kong/insomnia#3652] [Ideas]
**Is your feature request related to a problem? Please describe.**
I have different projects/services that need to be accessed with different HTTP proxy settings.  Right now when I switch between them I have to change the Collection (Workspace) *and* edit "HTTP Network Proxy" in the global Insomnia Preferences.

**Describe the solution you'd like**
It would be great If I could assign (or override) separate HTTP(S) proxies to each Collection.  That would make switching between projects a breeze.

**Describe alternatives you've considered**
Manual switching works, it's just tiring and easy to forget (leading to frustration when I realize that's what's causing things to not work).

**Additional context**
* There was [a mention](https://github.com/Kong/insomnia/issues/215#issuecomment-311138297) of adding per-Workspace settings (including proxies), but it was about 4 years ago now.
* There was also a [similar request](https://github.com/Kong/insomnia/issues/1385), but for per-request settings.  That would work for my use case, but I don't need that level of granularity (per-Collection or per-Environment would be easier for me to configure and would be flexible enough for my use case).

Thanks for such a useful product and for taking the time to have a look at this issue!

**Comment by dimitropoulos** [comment]
where would you think it best to control a setting like this.  currently I don't think we have a direct comparison or place to put this (maybe Collection Settings is best? but that's, still, a little hidden (maybe that's ok?)).

**Reply by katylava** [reply]
Would it make sense as an environment setting instead of a colllection setting? This would work better in my situation, fwiw.

**Reply by dimitropoulos** [reply]
Could you say more about what you mean by "environment setting"?  (or give an example)

**Reply by asyba** [reply]
@dimitropoulos like this:
![image](https://user-images.githubusercontent.com/1466530/131534196-dd59ce09-eca2-4d92-a85f-465e3cbb38fc.png)

**Reply by katylava** [reply]
I just realized it doesn't make much sense because the environment JSON defines variables. If you add proxy settings to it then you are mixing uses and that's confusing. You'd have to organize the JSON differently to make it not confusing, and that seems like a big change.

<img width="1245" alt="katy_s_Personal_Project_-_proxy_example_–__helloworld" src="https://user-images.githubusercontent.com/10575/131535017-6fc1499b-0f37-492c-b206-3b076bfa7760.png">

**Reply by asyba** [reply]
for me is not confused, Insomnia should create some variables that can be override to change the proxy on environment JSON.

**Reply by katylava** [reply]
As for the actual use case -- Akamai provides a staging environment that you can proxy through. So the ideal workflow is:

1. in akamai, make configuration changes and activate on "staging"
2. select "akamai staging" environment in insomnia
2. make request from insomnia make sure response is what is expected
3. in akamai, activate updated configuration on the real domain
4. switch environment to "qa" in insomnia
5. make same request as before to make sure it's still what you expect

**Reply by dimitropoulos** [reply]
Thanks.  I wanted to make sure we're talking about the same thing.  There are no "environment settings".  We have environment variables and we contain those in environments (you can switch between) but aside from assigning a name and a color there're no more settings on environments.

And on that note,

> I just realized it doesn't make much sense

Yes, that's what I am thinking as well.  Sure, it could be conceivably possible for the insomnia application itself to look at environment variables and change the behavior of the application, but there's no precedent for this and I don't see us ever doing it.  The reason is that environment variables are solely in the "userland" side, while we configure the application with settings in the app preferences.

**Reply by katylava** [reply]
out of curiosity though, why do proxy settings have to be part of application config instead of living in userland? i mean, is there a technical reason, or is it because that’s just how insomnia is currently programmed and it would be hard to change?

**Reply by dimitropoulos** [reply]
it's not that it'd be hard to change, but just that there's a architectural divide in play.  environments are used for feeding data into requests - like templates, and not configuring the application.  We have a separate place for configuring the applications (preferences).  Are we talking about the same things?

**Reply by katylava** [reply]
i probably just don’t know what i’m talking about — i don’t know why proxy is an application config and not a request config. could insomnia do something like curl’s `--resolve` option per-request instead? that’s how i send a request through akamai staging when i’m using curl. honestly i don’t know the difference between a proxy and whatever `--resolve` does, so please forgive me if this sounds dumb.

**Reply by develohpanda** [reply]
Proxy is an application setting (currently) but is used at a request level (to my knowledge). It would make sense to allow proxy settings to be overridden for a particular request (or folder or collection or project), the same way folder environments override sub environments, which in turn override the base environment.

I don't think there is any significant reason for why proxy settings are at an application level rather than request level. I think it logically makes sense to set proxy settings to any level and apply to everything downstream unless there's an override, but at the very least allow for overriding at a request level.

**Comment by Shackelford-Arden** [comment]
Glad I'm not the only one who thought about this in some way 😄 

After reading through the current thread, a couple ideas come to mind:

- Could we get the proxy setting added or included to the `Collection Settings`?
  - Seems there are already some type of collection settings since we can set the Client Certificates
- Could we get a simpler way to toggle the proxy (even if it affects the application globally) from within a collection?

**Reply by katylava** [reply]
FWIW, as a collection setting it wouldn't work as well for the Akamai staging use case, since I want to make the same requests through the proxy as I do without the proxy.

**Comment by UltimateFighter** [comment]
How to proceed with this?

**Comment by jacob-winkler** [comment]
Something to solve this would be very helpful. Start with the simplest case, make it a request setting. See screenshot.
![image](https://github.com/Kong/insomnia/assets/26692191/80181604-a046-4aa7-bc94-8055d107ecc1)
